Move user input logic into servoshell

atbrakhi opened the pull request:
We have user input logic spread across different functions and files. This PR brings all the logic and tests to `servoshell`. See comment https://github.com/servo/servo/pull/30148#pullrequestreview-1588405617


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

atbrakhi committed:
cleanup and move user input logix into servoshell

atbrakhi committed:
fix fmt

atbrakhi commented on ports/servoshell/Cargo.toml:
```diff
@@ -57,7 +57,7 @@ glow = "0.12.2"
 lazy_static = { workspace = true }
 libc = { workspace = true }
-libservo = { path = "../../components/servo" }
+libservo = { path = "../../components/servo", features=["media-gstreamer"] }
```
This is because of https://github.com/servo/servo/issues/30189

mrobinson replied:
Hrm...I think this will need to be addressed before this can land. I'm pretty sure we don't want to unconditionally enable this feature.

atbrakhi replied:
Removed this change from this PR, as issue was resolved in another PR

mrobinson commented on ports/servoshell/tests/mod.rs:
```diff
@@ -0,0 +1,6 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#[cfg(test)]
+mod servoshell_test;
```
I think you can just include the contents of `servoshell_test.rs` here.

atbrakhi replied:
done in [fab229d](https://github.com/servo/servo/pull/30238/commits/fab229d33f3b9003c8351cfae11a5c22187a2254)

mrobinson commented on components/config/opts.rs:
```diff
@@ -25,7 +25,7 @@ pub struct Opts {
     pub is_running_problem_test: bool,
 
     /// The initial URL to load.
-    pub url: Option<ServoUrl>,
+    pub url: Option<String>,
```
It feels like if we are moving the parsing out of opt.rs, we should also move the command-line argument out of `opt.rs`. `servoshell` has a few command-line arguments...perhaps this can move there.

atbrakhi replied:
I am not sure if [7214d3a](https://github.com/servo/servo/pull/30238/commits/7214d3a19dd2c87cfd5f969f1db5b83bff30fe61) is what you meant. But I realized we are only using opt url in this function so I have moved it all inside `get_default_url()`

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

atbrakhi committed:
moves test from servoshell file

atbrakhi committed:
move command-line args into servoshell

atbrakhi committed:
remove feature media-gstreamer

atbrakhi committed:
fix fmt

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

atbrakhi committed:
move user input logic code into lib to make it more testable

atbrakhi commented:
There are still a few things that needs to be resolved in this PR. Once they are resolved, I will mark it as ready for review.

atbrakhi committed:
remove opts_matches in fn instead get it from main2

atbrakhi committed:
remove pub and fix import

atbrakhi committed:
add licence in new file

atbrakhi commented:
New changes:
- Taking inspiration from `mbrubeck` suggestion [here](https://users.rust-lang.org/t/unit-tests-for-bin-crate/54403) on writing unit test in bin crate, I have moved the functions we are trying to test in `lib.rs`. It made it easier to call the functions in tests and other files. I am all ears if you think this might not be the best
- Moved command-line argument out of `opt.rs`

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

atbrakhi committed:
revert passing Matches, instead pass Option String

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

mrobinson commented on ports/servoshell/Cargo.toml:
```diff
@@ -13,6 +13,10 @@ name = "servo"
 path = "main.rs"
 bench = false
 
+[lib]
+name = "servoshell"
+path = "lib.rs"
+
```
I don't think it's strictly necessary to move the functions to a lib crate. It probably doesn't make sense to do this unless it is necessary (to write doctests as in the link you gave) or if we are going to be moving a lot of the logic from `servoshell` into a library.

atbrakhi committed:
review update, also move sanitize fn to parser file

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

atbrakhi committed:
fmt fix

atbrakhi commented:
@mrobinson thanks for the review again. I have updated the PR with the suggestions.

The only new thing in last commit is that I have moved `sanitize_url()` from `browser.rs` to `parser.rs`, so that we can have all the url parsing logic at one place.

mrobinson commented on ports/servoshell/parser.rs:
```diff
@@ -0,0 +1,62 @@
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::env;
+use std::path::Path;
+
+use log::warn;
```
Please remove the blank line here.

mrobinson commented on ports/servoshell/app.rs:
```diff
@@ -8,23 +8,22 @@ use crate::browser::Browser;
+use crate::parser::get_default_url;
 use crate::window_trait::WindowPortsMethods;
 use crate::{headed_window, headless_window};
+
```
Please remove this blank line in the middle of the use statements.

mrobinson reviewed (approved):
Thanks for the fixes! Just a couple nitpicks, but otherwise this looks good to me.

atbrakhi committed:
review fix: remove extra line

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.
